### PR TITLE
[rpi] Adds basic support for GPIO on Raspberry Pi boards

### PR DESCRIPTION
**Pull request by henrikssn**
The purpose of this is to simplify driver development, using bitbang SPI&I2C on the RPi.

It is essentially just a wrapper around WiringPi.

### CODE REVIEW

File: tools/devices/hosted/rpi.xml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<!DOCTYPE rca SYSTEM "../devicefile.dtd">
+<modm version="1.0">
+  <device platform="hosted" family="rpi">
+    <naming-schema>{platform}-{family}</naming-schema>
+
+    <attribute-core value="x86"/>
```
Comment (salkinium) [inline tools/devices/hosted/rpi.xml]: Is that even used anywhere by modm? I think it can be removed.

**Review by salkinium — DISMISSED** [review]
Nice job! I didn't expect so few changes to get this to work, but then again Linux is doing all the heavy lifting here.

File: src/modm/platform/gpio/rpi/module.lb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+# -----------------------------------------------------------------------------
+
+
+def init(module):
+    module.name = ":platform:gpio"
+    module.description = "Hosted GPIO for Raspberry Pi"
+
+
+def prepare(module, options):
+    if not options[":target"].has_driver("gpio"):
```
Comment (salkinium) [inline src/modm/platform/gpio/rpi/module.lb]:
```suggestion
    if not options[":target"].has_driver("gpio:rpi"):
```

Comment (henrikssn) [reply src/modm/platform/gpio/rpi/module.lb]: Done

File: tools/devices/hosted/rpi.xml (hunk 1 above)
Comment (henrikssn) [reply tools/devices/hosted/rpi.xml]: Removed

**Comment by henrikssn** [comment]
> Nice job! I didn't expect so few changes to get this to work, but then again Linux is doing all the heavy lifting here.

Thanks! Yeah, it is mostly Linux and WiringPi which solved the hard stuff, this is mostly a wrapper... Anyway, I tested the BitBangSPI implementation with this and that worked fine. Might be the easiest way to run hardware driver tests (if we ever get to setting up a CI for that)

File: src/modm/platform/gpio/rpi/module.lb (hunk 2 above)
Comment (salkinium) [reply src/modm/platform/gpio/rpi/module.lb]: Nope 😉

File: src/modm/platform/gpio/rpi/module.lb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+# -----------------------------------------------------------------------------
+
+
+def init(module):
+    module.name = ":platform:gpio"
+    module.description = "Hosted GPIO for Raspberry Pi"
+
+
+def prepare(module, options):
+<<<<<<< HEAD
```
Comment (rleh) [inline src/modm/platform/gpio/rpi/module.lb]: Nopenopenope

File: src/modm/board/raspberrypi_4/module.lb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+#
+# This file is part of the modm project.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+# -----------------------------------------------------------------------------
+
+def init(module):
+    module.name = ":board:raspberrypi-4"
```
Comment (rleh) [inline src/modm/board/raspberrypi_4/module.lb]: Why (only) Raspberry Pi 4? As far as I can see this should also work on Pi 1, 2 and 3.

File: src/modm/platform/gpio/rpi/module.lb (hunk 2 above)
Comment (henrikssn) [reply src/modm/platform/gpio/rpi/module.lb]: Meh, forgot to save the file in my editor :(

File: src/modm/platform/gpio/rpi/module.lb (hunk 3 above)
Comment (henrikssn) [reply src/modm/platform/gpio/rpi/module.lb]: Removed

File: src/modm/board/raspberrypi_4/module.lb (hunk 4 above)
Comment (henrikssn) [reply src/modm/board/raspberrypi_4/module.lb]: The pin mapping / peripherals is different so if we want to extend the board with that then they would need to be separate.

Obviously you can still use this board on a RPi 1/2/3 but then you are on your own if it breaks in the future :)

Comment (rleh) [reply src/modm/board/raspberrypi_4/module.lb]: But inside this pull request I don't see any specialization regarding the RPi 4, therefore I favour a generic name for this module.

**Comment by rleh** [comment]
https://circleci.com/gh/modm-io/modm/20845:
```
python3 tools/scripts/synchronize_docs.py -d

git diff
diff --git a/README.md b/README.md
index 79a9df804..eaf76f7d4 100644
--- a/README.md
+++ b/README.md
@@ -157,6 +157,7 @@ documentation.
 <td align="center">NUCLEO-L476RG</td>
 <td align="center">OLIMEXINO-STM32</td>
 </tr><tr>
+<td align="center">RASPBERRYPI-4</td>
 <td align="center">SAMD21-MINI</td>
 <td align="center">STM32F030F4P6-DEMO</td>
 </tr>

Please synchronize the modm documentation:

    $ python3 tools/scripts/synchronize_docs.py

and then commit the results!
```

File: src/modm/board/raspberrypi_4/module.lb (hunk 4 above)
Comment (henrikssn) [reply src/modm/board/raspberrypi_4/module.lb]: Renamed it to "raspberrypi"

**Review by rleh — APPROVED** [review]

**Comment by henrikssn** [comment]
Ready to merge

**Comment by rleh** [comment]
Could you squash and rebase your commits?

**Comment by salkinium** [comment]
I’ll rebase/merge it later.

**Comment by salkinium** [comment]
> I’ll rebase/merge it later.

Restructured commits and fixed a small refactoring bug, I'll merge once the CI passes.

**Pull request merged**